### PR TITLE
release 0.2.0-rc.1

### PR DESCRIPTION
**Pull request by matthewhanson**
**Related Issue(s):** #
#73 
#68 
#70
#69

**Description:**
This PR updates to PySTAC 1.0 along with:
- remove redundant code in Client class
- remove previously deprecated functions Client.collections and Client.items
- Adds CollectionClient class
- Adds reading from /collections endpoint to get all collections at once
- Redefine get_items for Catalogs and Collections to fetch items using the API endpoints instead of crawling (will fall back to crawling if non-conforming)
- Renames stac_io module to stac_api_io
- Adds stac_object_from_dict to stac_api_io module, allowing resolved Catalog and Collection STAC objects to be returned as Client or CollectionClient, respectively
- refactors conformance into StacApiIO
- add `ignore_conformance` to Catalog.open() to allow the use of any functions regardless of advertised conformance classes

**PR Checklist:**

- [x] Code is formatted
- [x] Tests pass
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac-api-client/blob/main/CHANGELOG.md)

### CODE REVIEW

File: pystac_client/stac_api_io.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+        uris = CONFORMANCE_URIS.get(conformance_class.name, None)
+
+        if uris is None:
+            raise Exception(f"Invalid conformance class {conformance_class}")
+
+        if not any(uri in uris for uri in self._conformance):
+            raise NotImplementedError(f"{conformance_class} not supported")
+
+        return True
```
Comment (gadomski) [inline pystac_client/stac_api_io.py:225]:
```suggestion
        return True
        
    def set_conformance(self, conformance: Optional[List[str]]) -> None:
        """Sets (or clears) the conformances for this StacIO."""
        self._conformance = conformance

```

File: pystac_client/stac_api_io.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        if self._conformance is None:
+            return True
+
+        uris = CONFORMANCE_URIS.get(conformance_class.name, None)
+
+        if uris is None:
+            raise Exception(f"Invalid conformance class {conformance_class}")
+
+        if not any(uri in uris for uri in self._conformance):
+            raise NotImplementedError(f"{conformance_class} not supported")
```
Comment (gadomski) [inline pystac_client/stac_api_io.py:223]: Is there a reason to raise an error instead of just returning `False`?

**Review by gadomski — CHANGES_REQUESTED** [review]

File: pystac_client/item_search.py [hunk 3]
```diff
@@ -355,7 +364,7 @@ def _format_fields(self, value: Optional[FieldsLike]) -> Optional[Fields]:
         if value is None:
             return None
 
-        self.conforms_to(ConformanceClasses.FIELDS)
+        self._stac_io.conforms_to(ConformanceClasses.FIELDS)
```
[the remark this answers is not in the document]
Comment (matthewhanson) [reply pystac_client/item_search.py:367]: Yes, probably that would make sense.  I guess alternately there could be a `conforms_to` that returns a bool and an `assert` variation that throws an error.

Related, I don't really like having a conforms_to function in both Client/StacApiIO (and CollectionClient/StacApiIO), suggestions welcome.

File: pystac_client/stac_api_io.py (hunk 2 above)
Comment (matthewhanson) [reply pystac_client/stac_api_io.py:223]: I implemented it as a bool, then added back the exception raising.

File: pystac_client/client.py [hunk 4]
```diff
@@ -80,75 +59,72 @@ def open(cls, url=None, headers=None):
             raise TypeError(
                 "'url' must be specified or the 'STAC_URL' environment variable must be set.")
 
-        stac_io = StacApiIO(headers=headers)
-
-        catalog = cls.from_file(url, stac_io)
-        return catalog
+        cat = cls.from_file(url, headers=headers)
+        if ignore_conformance:
+            cat._stac_io._conformance = None
```
[the remark this answers is not in the document]
Comment (matthewhanson) [reply pystac_client/client.py:64]: I think I've mixed approaches here and it all could use some cleanup.  In some cases it follows the conventions that PySTAC does and in others it doesn't.  But yes, getter/setter should be implemented here

File: pystac_client/cli.py [hunk 5]
```diff
@@ -139,8 +165,21 @@ def cli():
...
+        client = Client.open(url, headers=headers, ignore_conformance=ignore_conformance)
+    except Exception as e:
+        print(e)
+        return 1
+
     if cmd == 'search':
-        return search(**args)
+        return search(client, **args)
+    elif cmd == 'collections':
+        args.pop('limit')
```
[the remark this answers is not in the document]
Comment (matthewhanson) [reply pystac_client/cli.py:181]: This was a temporary hack. Limit should be respected even for /collections, but right now the spec doesn't allow for pagination at that endpoint.  I'm going to move the limit arg to be just for search

File: pystac_client/item_search.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gadomski) [reply pystac_client/item_search.py:367]: > Related, I don't really like having a conforms_to function in both Client/StacApiIO (and CollectionClient/StacApiIO), suggestions welcome.

Does the `Client.conforms_to` just pass down to the `StacApiIO.conforms_to`? If so, seems fine to me.

**Pull request merged**